File: src/org/cappuccino_project/ide/intellij/plugin/psi/utils/ObjJHasContainingClassPsiUtil.java
package org.cappuccino_project.ide.intellij.plugin.psi.utils;

import com.intellij.psi.PsiElement;
import org.cappuccino_project.ide.intellij.plugin.psi.ObjJImplementationDeclaration;
import org.cappuccino_project.ide.intellij.plugin.psi.ObjJMethodHeader;
import org.cappuccino_project.ide.intellij.plugin.psi.ObjJSelector;
import org.cappuccino_project.ide.intellij.plugin.psi.ObjJSelectorLiteral;
import org.cappuccino_project.ide.intellij.plugin.psi.interfaces.ObjJClassDeclarationElement;
import org.cappuccino_project.ide.intellij.plugin.psi.interfaces.ObjJCompositeElement;
import org.cappuccino_project.ide.intellij.plugin.psi.interfaces.ObjJHasContainingClass;
import org.cappuccino_project.ide.intellij.plugin.psi.types.ObjJClassType;
import org.cappuccino_project.ide.intellij.plugin.stubs.interfaces.ObjJSelectorLiteralStub;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

import static org.cappuccino_project.ide.intellij.plugin.psi.utils.ObjJMethodCallPsiUtil.isUniversalMethodCaller;

public class ObjJHasContainingClassPsiUtil {

    public static boolean isSimilarClass(@NotNull final String methodContainingClass, @NotNull final String className) {
        return isUniversalMethodCaller(className) || isUniversalMethodCaller(methodContainingClass) || Objects.equals(className, methodContainingClass);
    }

    @Nullable
    public static ObjJClassDeclarationElement getContainingClass(@Nullable
                                                                         PsiElement element) {
        return ObjJTreeUtil.getParentOfType(element, ObjJClassDeclarationElement.class);
    }

    @NotNull
    public static String getContainingClassName(ObjJMethodHeader methodHeader) {
        if (methodHeader.getStub() != null) {
            return methodHeader.getStub().getContainingClassName();
        }
        ObjJClassDeclarationElement containingClass = methodHeader.getContainingClass();
        return containingClass != null ? containingClass.getClassType().getClassName() : ObjJClassType.UNDEF_CLASS_NAME;
    }

    @NotNull
    public static String getContainingClassName(ObjJCompositeElement compositeElement) {
        final ObjJClassDeclarationElement classDeclarationElement = getContainingClass(compositeElement);
        if (classDeclarationElement == null) {
            return ObjJClassType.UNDEF.getClassName();
        }
        return classDeclarationElement.getClassNameString();
    }

    @NotNull
    public static String getContainingClassName(@Nullable ObjJClassDeclarationElement classDeclarationElement) {
        return classDeclarationElement != null ? classDeclarationElement.getClassNameString() : ObjJClassType.UNDEF_CLASS_NAME;
    }

    @NotNull
    public static String getContainingClassName(@Nullable
                                                        ObjJSelectorLiteral selectorLiteral) {
        if (selectorLiteral == null) {
            return ObjJClassType.UNDEF_CLASS_NAME;
        }
        ObjJSelectorLiteralStub stub = selectorLiteral.getStub();
        if (stub != null) {
            return stub.getContainingClassName();
        }
        return getContainingClassName(selectorLiteral.getContainingClass());
    }
    @NotNull
    public static List<String> getContainingClassNamesFromSelector(@NotNull List<ObjJSelector> elements) {
        List<String> out = new ArrayList<>();
        for (ObjJSelector element : elements) {

        }
        return out;
    }

    @NotNull
    public static List<String> getContainingClassNames(@NotNull List<PsiElement> elements) {
        List<String> out = new ArrayList<>();
        for (PsiElement element : elements) {
            String className = null;
            if (elements instanceof ObjJHasContainingClass) {
                className = ((ObjJHasContainingClass)element).getContainingClassName();
            } else if (element instanceof  ObjJCompositeElement){
                className = getContainingClassName((ObjJCompositeElement)element);
            }
            if (className != null && !out.contains(className)) {
                out.add(className);
            }
        }
        return out;
    }

    public static boolean sharesContainingClass(@NotNull String className, @NotNull PsiElement... elements) {
        for (PsiElement element : elements) {
            if (!ObjJHasContainingClassPsiUtil.hasContainingClass(element, className)) {
                return false;
            }
        }
        return true;
    }

    public static boolean hasContainingClass(@Nullable PsiElement element, @NotNull String className) {
        return  className.equals(ObjJClassType.UNDETERMINED) ||
                (element != null && element instanceof ObjJHasContainingClass && ((ObjJHasContainingClass) element).getContainingClassName().equals(className));
    }



    public static String getContainingClassOrFileName(PsiElement psiElement) {
        String containingClassName = null;
        if (psiElement instanceof ObjJHasContainingClass) {
            containingClassName = ((ObjJHasContainingClass)psiElement).getContainingClassName();
        }
        if (containingClassName == null) {
            ObjJClassDeclarationElement classDeclarationElement = ObjJTreeUtil.getParentOfType(psiElement, ObjJClassDeclarationElement.class);
            if (classDeclarationElement != null) {
                containingClassName = classDeclarationElement.getClassNameString();
            }
        }
        if (containingClassName == null) {
            containingClassName = psiElement.getContainingFile().getVirtualFile().getName();
        }
        return containingClassName;
    }

    public static String getContainingSuperClassName(@NotNull ObjJCompositeElement element) {
        ObjJClassDeclarationElement classDeclarationElement = getContainingClass(element);
        if (classDeclarationElement == null || !(classDeclarationElement instanceof ObjJImplementationDeclaration)) {
            return null;
        }
        return ObjJClassDeclarationPsiUtil.getSuperClassName((ObjJImplementationDeclaration)classDeclarationElement);
    }


}

File: src/org/cappuccino_project/ide/intellij/plugin/psi/utils/ObjJHasContainingClassPsiUtil.kt
package org.cappuccino_project.ide.intellij.plugin.psi.utils

import com.intellij.psi.PsiElement
import org.cappuccino_project.ide.intellij.plugin.psi.ObjJImplementationDeclaration
import org.cappuccino_project.ide.intellij.plugin.psi.ObjJMethodHeader
import org.cappuccino_project.ide.intellij.plugin.psi.ObjJSelector
import org.cappuccino_project.ide.intellij.plugin.psi.ObjJSelectorLiteral
import org.cappuccino_project.ide.intellij.plugin.psi.interfaces.ObjJClassDeclarationElement
import org.cappuccino_project.ide.intellij.plugin.psi.interfaces.ObjJCompositeElement
import org.cappuccino_project.ide.intellij.plugin.psi.interfaces.ObjJHasContainingClass
import org.cappuccino_project.ide.intellij.plugin.psi.types.ObjJClassType

import java.util.ArrayList


object ObjJHasContainingClassPsiUtil {

    fun isSimilarClass(methodContainingClass: String, className: String): Boolean {
        return isUniversalMethodCaller(className) || isUniversalMethodCaller(methodContainingClass) || className == methodContainingClass
    }

    fun getContainingClass(element: PsiElement?): ObjJClassDeclarationElement<*>? {
        return element.getParentOfType( ObjJClassDeclarationElement::class.java)
    }

    fun getContainingClassName(methodHeader: ObjJMethodHeader): String {
        val stub = methodHeader.stub
        if (stub != null) {
            return stub.containingClassName
        }
        val containingClass = methodHeader.containingClass
        return if (containingClass != null) containingClass.classType.className else ObjJClassType.UNDEF_CLASS_NAME
    }

    fun getContainingClassName(compositeElement: ObjJCompositeElement): String {
        val classDeclarationElement = getContainingClass(compositeElement) ?: return ObjJClassType.UNDEF.className
        return classDeclarationElement.classNameString
    }

    fun getContainingClassName(classDeclarationElement: ObjJClassDeclarationElement<*>?): String {
        return classDeclarationElement?.classNameString ?: ObjJClassType.UNDEF_CLASS_NAME
    }

    fun getContainingClassName(selectorLiteral: ObjJSelectorLiteral?): String {
        if (selectorLiteral == null) {
            return ObjJClassType.UNDEF_CLASS_NAME
        }
        val stub = selectorLiteral.stub
        return stub?.containingClassName ?: getContainingClassName(selectorLiteral.containingClass)
    }

    fun getContainingClassNamesFromSelector(elements: List<ObjJSelector>): List<String> {
        val out = ArrayList<String>()
        for (element in elements) {

        }
        return out
    }

    fun getContainingClassNames(elements: List<PsiElement>): List<String> {
        val out = ArrayList<String>()
        for (element in elements) {
            var className: String? = null
            if (elements is ObjJHasContainingClass) {
                className = (element as ObjJHasContainingClass).containingClassName
            } else if (element is ObjJCompositeElement) {
                className = getContainingClassName(element)
            }
            if (className != null && !out.contains(className)) {
                out.add(className)
            }
        }
        return out
    }

    fun sharesContainingClass(className: String, vararg elements: PsiElement): Boolean {
        for (element in elements) {
            if (!ObjJHasContainingClassPsiUtil.hasContainingClass(element, className)) {
                return false
            }
        }
        return true
    }

    fun hasContainingClass(element: PsiElement?, className: String): Boolean {
        return className == ObjJClassType.UNDETERMINED || element != null && element is ObjJHasContainingClass && element.containingClassName == className
    }


    fun getContainingClassOrFileName(psiElement: PsiElement): String {
        var containingClassName: String? = null
        if (psiElement is ObjJHasContainingClass) {
            containingClassName = psiElement.containingClassName
        }
        if (containingClassName == null) {
            val classDeclarationElement = psiElement.getParentOfType( ObjJClassDeclarationElement::class.java)
            if (classDeclarationElement != null) {
                containingClassName = classDeclarationElement.classNameString
            }
        }
        if (containingClassName == null) {
            containingClassName = psiElement.containingFile.virtualFile.name
        }
        return containingClassName
    }

    fun getContainingSuperClassName(element: ObjJCompositeElement): String? {
        val classDeclarationElement = getContainingClass(element)
        return if (classDeclarationElement == null || classDeclarationElement !is ObjJImplementationDeclaration) {
            null
        } else classDeclarationElement.getSuperClassName()
    }


}
